Gen component collection and help

deinhofer opened the pull request:
An MD5 sum is calculated out of all plugin-jars in the bin/ARE directory. The hash is stored in the file tmp/caching/pluginJarsHash.txt

The next startup, the newly calculated MD5 hash is compared to the stored value in the file. If it changed or any of the cache files (profile/loader_componentlist.ini, data/componentCollections/defaultComponentCollection.abd) is missing, the cache files are regenerated.

klues commented on ARE/middleware/src/main/java/eu/asterics/mw/are/AsapiSupport.java:
```diff
@@ -216,39 +190,25 @@ public String getComponentDescriptorsAsXml() throws AREAsapiException {
-                                        .getResource(DefaultBundleModelParser.BUNDLE_DESCRIPTOR_RELATIVE_URI);
-                                if (bundleDescriptorURL != null) {
-                                    try {
-                                        response += getFormattedBundleDescriptorStringOfComponentTypeId(
```
why this formatting is not necessary anymore? (`getFormattedBundleDescriptorStringOfComponentTypeId()`)

deinhofer replied:
it has moved to https://github.com/asterics/AsTeRICS/blob/gen-component-collection-and-help/ARE/middleware/src/main/java/eu/asterics/mw/are/BundleManager.java#L747

klues replied:
ah, ok. thx

klues commented on ARE/middleware/src/main/java/eu/asterics/mw/are/AsapiSupport.java:
```diff
@@ -381,29 +341,9 @@ public Object call() throws Exception {
-                                    }
-                                }
-                            }                            
+                            res.add(getComponentDescriptorsAsXml());
```
so whats the difference between this method `getBundelDescriptors()` and the other one `getComponentDescriptorsAsXml()` if you are returning the result of `getComponentDescriptorsAsXml()` here? Why not remove one of the two methods?

klues replied:
by the way: `getBundelDescriptors()` has a typo, shouldnt it be `getBundleDescriptors()`? So maybe we can change it...

deinhofer replied:
unfortunately both methods are used, although the functionality is more a less equivalent. One is used by the C# ACS and the other one is used by the WebACS. I actually already did a hack by just adding the contents of the whole component collection to the first element of the List<String> collection in getBundleDescriptors.

Yes fixing the typo makes sense.

klues commented on ARE/middleware/src/main/java/eu/asterics/mw/are/AsapiSupport.java:
```diff
@@ -171,36 +175,6 @@ public AsapiSupport() {
-
-    /**
      * Returns an xml String containing the component collection (bundle
      * descriptors of every AsTeRiCS component). This function searches in the
```
"This function searches in the bin/ARE folder to discover the created components." is not correct anymore, is it? I think now we are just returning the cache file?!

deinhofer replied:
Yes, will fix this

klues commented on ARE/middleware/src/main/java/eu/asterics/mw/services/ResourceRegistry.java:
```diff
@@ -622,8 +629,12 @@ public static String getResourceContentAsString(InputStream inputStream) throws
+        try {
+            String modelAsString=IOUtils.toString(new BOMInputStream(inputStream),ResourceRegistry.ARE_FILE_ENCODING);       
+            return modelAsString;
+        }finally {
```
this line is not formatted

klues commented on ARE/middleware/src/main/java/eu/asterics/mw/services/ResourceRegistry.java:
```diff
@@ -638,20 +649,52 @@ public static String getResourceContentAsString(InputStream inputStream) throws
+    }
+    
+    /**
+     * Stores the given data to the given OutputStream and encoded as {@link ResourceRegistry#ARE_FILE_ENCODING}. 
```
comment does not mention that output stream is closed.

deinhofer replied:
ok

klues commented on ARE/services/WebService/src/main/java/eu/asterics/mw/webservice/SseResource.java:
```diff
@@ -100,10 +100,11 @@ public static void initializeDataChannelListeners() {
         // initialize accordint to the new model
-        for (IChannel channel : DeploymentManager.instance.getCurrentRuntimeModel().getChannels()) {
-            dataChannelBroadcaster.put(channel.getChannelID(), new HashSet<EventOutput>());
+        if(DeploymentManager.instance.getCurrentRuntimeModel()!=null) {
```
this line is not formatted

klues commented on ARE/middleware/src/main/java/eu/asterics/mw/are/BundleManager.java:
```diff
@@ -132,16 +144,173 @@ public BundleManager(ModelValidator modelValidator) {
+        	LOADER_COMPONENTLIST_CACHE_FILE_URI=ResourceRegistry.getInstance().getResource(LOADER_COMPONENTLIST_LOCATION,RES_TYPE.PROFILE);
+        	
+        	if(ResourceRegistry.getInstance().isOSGIMode()) {
+        	    String md5Sum=generateMD5Sum();
```
please format

klues commented on ARE/middleware/src/main/java/eu/asterics/mw/are/BundleManager.java:
```diff
@@ -132,16 +144,173 @@ public BundleManager(ModelValidator modelValidator) {
+                byte[] readBytes = new byte[1024];
+                int numRead=0;
+                while((numRead=dis.read(readBytes))!=-1) {
+                    MDInstance.update(readBytes,0,numRead);
```
I do not think you need to update the `MDInstance` here. You are using a `DigestInputStream` that does this automatically. Actually I think this way you are calculating a wrong MD5 hash because all bytes are added two times to the MessageDigest.

klues replied:
instead of calculating the hash "manually" here you could also use:
https://commons.apache.org/proper/commons-codec/apidocs/org/apache/commons/codec/digest/DigestUtils.html#digestAsHex-java.io.InputStream-
I think you added the commons-codec package with this merge?!

deinhofer replied:
regarding calculation: you are right, it also works without calling update

I did not want to use digestAsHex here because I wanted to add all bytes of all plugin jars to one instance of DigestInputStream and there is no method for such a case.

klues commented on ARE/middleware/src/main/java/eu/asterics/mw/are/BundleManager.java:
```diff
@@ -132,16 +144,173 @@ public BundleManager(ModelValidator modelValidator) {
+    private String generateMD5Sum() throws NoSuchAlgorithmException, MalformedURLException, IOException {
+        logger.fine("Generating MD5 sum of bundle jars....");
+        long timeStart=System.currentTimeMillis();
+        MessageDigest MDInstance=MessageDigest.getInstance("MD5");
```
`MDInstance` is a strange name, I think it should start with a small letter, like all variables. So e.g. `mdInstance`?

deinhofer replied:
ok

klues commented on ARE/middleware/src/main/java/eu/asterics/mw/are/BundleManager.java:
```diff
@@ -132,16 +144,173 @@ public BundleManager(ModelValidator modelValidator) {
+     * @throws URISyntaxException
+     * @throws IOException
+     */
+    private boolean bundleChangeDetected(String md5Sum) throws URISyntaxException, IOException {
```
shouldnt `URISyntaxException` and `IOException` be catched and 'true' returned in this case?

klues commented on ARE/middleware/src/main/java/eu/asterics/mw/are/BundleManager.java:
```diff
@@ -132,16 +144,173 @@ public BundleManager(ModelValidator modelValidator) {
+     */
+    private void storeMD5Sum(String md5Sum) throws URISyntaxException, IOException {
+        logger.fine("Storing MD5 sum...");
+        URI pluginJarsHashFileURI=ResourceRegistry.getInstance().getResource(CACHING_PLUGIN_JARS_HASH_TXT, RES_TYPE.TMP);
```
why not using `ResourceRegistry#storeResource(java.lang.String, java.lang.String, ResourceRegistry.RES_TYPE)` here?

deinhofer replied:
ok

klues commented on ARE/middleware/src/main/java/eu/asterics/mw/are/BundleManager.java:
```diff
@@ -647,84 +805,57 @@ public void installServices() {
                 }
             }
             writer.flush();
             writer.close();
```
i think you do not need these lines anymore because you used an try-with-resource statement which does closing automatically, see https://docs.oracle.com/javase/7/docs/technotes/guides/language/try-with-resources.html

deinhofer replied:
correct

deinhofer commented:
* Fixed recommendations from @klues 

* made the set of bundle jars configurable in areProperties by default all bundle jars are used as base for the cache files generation. 
The behaviour can be configured with the property:
ARE.generate.caches.from.installable.bundles=true|false

true: The bundles are tried to install with OSGI for the current platform
false: The bundles are not tried to install

* moved location of profile/loader_componenlist.ini to tmp/caching/loader_componentlist.ini
* made message digest algorithm configurable in source code (MD5 or SHA-256). Default is still MD5 because it needed half of time of SHA-256 for 190 bundles 
* Improved AREProperties class to allow the storing of property associated comments
